### PR TITLE
Story/11608 fault tolerant name handling/11682 certificate person matching

### PR DESCRIPTION
**Pull request by ArturFriesen**
## Description
Add func to check if a person is the same. Based on the new matching rules.

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-11682

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift [hunk 1]
```diff
@@ -168,6 +168,10 @@ final class HealthCertificate: Codable, Equatable, Comparable, RecycleBinIdentif
 		digitalCovidCertificate.dateOfBirth
 	}
 
+	var trimmedDateOfBirth: String {
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift:171]:
```suggestion
	private var trimmedDateOfBirth: String {
```

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift [hunk 2]
```diff
@@ -280,6 +284,19 @@ final class HealthCertificate: Codable, Equatable, Comparable, RecycleBinIdentif
 			return hashData == $0.hash
 		}
 	}
+	
+	func belongsToSamePerson(_ other: HealthCertificate) -> Bool {
+		let hasSameDateOfBirth = self.trimmedDateOfBirth == other.trimmedDateOfBirth
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift:289]: As a performance optimization this could be a guard and we could only do the other operations if this is true. Just an idea!

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift [hunk 3]
```diff
@@ -280,6 +284,19 @@ final class HealthCertificate: Codable, Equatable, Comparable, RecycleBinIdentif
...
+		
+		let givenNameCompontents = Set<String>(self.name.givenNameGroupingComponents ?? [])
+		let otherGivenNameCompontents = other.name.givenNameGroupingComponents ?? []
+		let hasGivenNameIntersection = givenNameCompontents.intersection(otherGivenNameCompontents).isNotEmpty
+		
+		let familyNameComponents = Set<String>(self.name.familyNameGroupingComponents)
+		let hastFamilyNameIntersection = familyNameComponents.intersection(other.name.familyNameGroupingComponents).isNotEmpty
+		
+		return hasSameDateOfBirth && hasGivenNameIntersection && hastFamilyNameIntersection
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift:298]:
```suggestion
		let hasFamilyNameIntersection = familyNameComponents.intersection(other.name.familyNameGroupingComponents).isNotEmpty
		
		return hasSameDateOfBirth && hasGivenNameIntersection && hasFamilyNameIntersection
```

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift [hunk 4]
```diff
@@ -280,6 +284,19 @@ final class HealthCertificate: Codable, Equatable, Comparable, RecycleBinIdentif
 			return hashData == $0.hash
 		}
 	}
+	
+	func belongsToSamePerson(_ other: HealthCertificate) -> Bool {
+		let hasSameDateOfBirth = self.trimmedDateOfBirth == other.trimmedDateOfBirth
+		
+		let givenNameCompontents = Set<String>(self.name.givenNameGroupingComponents ?? [])
+		let otherGivenNameCompontents = other.name.givenNameGroupingComponents ?? []
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift:292]: We could move the fallback to `[]` into `givenNameGroupingComponents`

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/Extensions/Name+Extension.swift [hunk 5]
```diff
@@ -66,6 +74,35 @@ extension Name {
...
+		let components: [String] =
+		self.trimmingCharacters(in: .whitespaces)
+		// the string shall be trimmed for leading and trailing `<`
+			.trimmingCharacters(in: CharacterSet(charactersIn: "<"))
+		// any whitespace in the string shall be replaced by `<`
+			.replacingOccurrences(of: "\\s+", with: "<", options: .regularExpression)
+		// any occurence of more than one `<` shall be replaced by a single `<`
+			.replacingOccurrences(of: "<+", with: "<", options: .regularExpression)
+			.replacingOccurrences(of: "-", with: "<")
+			.replacingOccurrences(of: ".", with: "<")
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/Extensions/Name+Extension.swift:90]: Would it make sense to do these two first, before replacing multiple `<` with a single `<`? And maybe mention these steps in a comment as well

**Review by nickguendling — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/HealthCertificate+BelongsToSamePersonTests.swift [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+		}
+		
+		let description: String
+		let actHolderA: Holder
+		let actHolderB: Holder
+		let expIsSameHolder: Bool
+	}
+	
+	let testCases = [
+		TestCase(
```
Comment (mlenkeit) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/HealthCertificate+BelongsToSamePersonTests.swift:57]: Can you dynamically read this from JSON?

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/Name+ExtensionTests.swift [hunk 7]
```diff
@@ -54,5 +54,164 @@ class Name_ExtensionTests: XCTestCase {
...
+	}
+	
+	struct TestCase {
+		let description: String
+		let actName: String
+		let expectedNameComponents: [String]
+	}
+	
+	let testCases = [
+		TestCase(
```
Comment (mlenkeit) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/Name+ExtensionTests.swift:72]: Can you dynamically read this from JSON?

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/HealthCertificate+BelongsToSamePersonTests.swift (hunk 6 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/HealthCertificate+BelongsToSamePersonTests.swift:57]: @mlenkeit  There was only a yaml, when I was writing the tests.

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/Extensions/Name+Extension.swift (hunk 5 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/Extensions/Name+Extension.swift:89]: done

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift (hunk 1 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift]: done

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift (hunk 2 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift]: done 🏎💨

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift (hunk 4 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift]: done

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift (hunk 3 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift]: done

**Review by nickguendling — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/HealthCertificate+BelongsToSamePersonTests.swift (hunk 6 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/HealthCertificate+BelongsToSamePersonTests.swift]: Added dynamic test case loading.
Discovered unimplemented edge cases.
I like this test driven thingy ;)

**Review by naveeddotio — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/Extensions/Name+Extension.swift [hunk 8]
```diff
@@ -34,6 +34,14 @@ extension Name {
 			.trimmingCharacters(in: .whitespaces)
 	}
 	
+	var givenNameGroupingComponents: [String] {
+		return standardizedGivenName?.groupingComponents() ?? []
+	}
+	
+	var familyNameGroupingComponents: [String] {
+		return standardizedFamilyName.groupingComponents()
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/Extensions/Name+Extension.swift:42]: Can it also return empty array like line 38?

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/Extensions/Name+Extension.swift [hunk 9]
```diff
@@ -34,6 +34,14 @@ extension Name {
 			.trimmingCharacters(in: .whitespaces)
 	}
 	
+	var givenNameGroupingComponents: [String] {
+		return standardizedGivenName?.groupingComponents() ?? []
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/Extensions/Name+Extension.swift:38]: Do we need to return an empty array here?

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/HealthCertificate+BelongsToSamePersonTests.swift [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+						dateOfBirth: testCase.actHolderB.dob
+					)
+				),
+				validityState: .valid
+			)
+			
+			XCTAssertEqual(lhsCertificate.belongsToSamePerson(rhsCertificate), testCase.expIsSameHolder, "Test failed. Description: \(testCase.description)")
+		}
+	}
+
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/HealthCertificate+BelongsToSamePersonTests.swift:51]: Mark Private

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/Name+ExtensionTests.swift [hunk 11]
```diff
@@ -54,5 +54,29 @@ class Name_ExtensionTests: XCTestCase {
...
+				  return
+		}
+		
+		for testCase in testCases {
+			let name = Name(familyName: "", givenName: "", standardizedFamilyName: testCase.actName, standardizedGivenName: "")
+			XCTAssertEqual(name.familyNameGroupingComponents, testCase.expNameComponents, "Failed test case: \(testCase.description	)")
+		}
+	}
+	
+	struct TestCases: Decodable {
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/Name+ExtensionTests.swift:73]: Could be private as well

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/HealthCertificate+BelongsToSamePersonTests.swift (hunk 10 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/HealthCertificate+BelongsToSamePersonTests.swift:51]: @naveeddotio What exactly?

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/Name+ExtensionTests.swift (hunk 11 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/__tests__/Name+ExtensionTests.swift:73]: done
